File: C/Data_Structures/Trees/6_AVL_Tree/01_insertion.c
#include "assert.h"
#include "tree.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

// ── Tree operations ─────────────────────────────────────────────────────────
// initialize node
TreeNode *initNode(int data) {
  TreeNode *newnode = malloc(sizeof(*newnode));
  if (!newnode)
    perror("Memory allocation failed!\n");
  newnode->data = data;
  newnode->left = NULL;
  newnode->right = NULL;
  newnode->height = 1;

  return newnode;
}

// ── AVL helpers ────────────────────────────────────────────────────────────
// get height
int getHeight(const TreeNode *node) {
  return (node == NULL) ? 0 : node->height;
}

// update height
void updateHeight(TreeNode *node) {
  node->height = 1 + max(getHeight(node->right), getHeight(node->left));
}

// get maximum height
int max(int a, int b) { return (a > b) ? a : b; }

// get balance factor
int getBalanceFactor(TreeNode *node) {
  return (node == NULL) ? 0 : getHeight(node->left) - getHeight(node->right);
}

// ── AVL operations ─────────────────────────────────────────────────────────
// left rotation
// Before  (RR imbalance)              After Left rotaion
//      x                                y
//       \                             /  \
//        y                           x    T3
//       / \                         / \
//      T2  T3                      T1  T2
TreeNode *leftRotate(TreeNode *x) {
  TreeNode *y = x->right;
  TreeNode *T3 = y->right;

  // perform rotation
  y->left = x;
  x->right = T3;

  // update height
  updateHeight(y);
  updateHeight(x);

  // return new root
  return y;
}

// right rotation
// Before (LL imbalance)          After Right Rotation
//     y (height=3, BF=+2)           x (new root, BF=0 or ±1)
//    /                           /   \
//   x                           T1    y
//  / \                               / \
// T1  T2                            T2  T3
TreeNode *rightRotate(TreeNode *y) {
  TreeNode *x = y->left;
  TreeNode *T2 = x->right;

  // perform rotation
  y->left = T2;
  x->right = y;

  // update height
  updateHeight(y);
  updateHeight(x);

  // return new root
  return x;
}

// ── insert node ──────────────────────────────────────────────────────────────
// insert node in tree
TreeNode *insertNode(TreeNode *root, int data) {
  // NOTE : insert node
  if (!root)
    return initNode(data);

  if (data <= root->data) {
    root->left = insertNode(root->left, data);
  } else {
    root->right = insertNode(root->right, data);
  }

  // NOTE :  update height
  updateHeight(root);

  // NOTE : get balance factor
  int balance_factor = getBalanceFactor(root);
  printf("Balance factor of %d: %d\n", root->data, balance_factor);

  // NOTE : four rotation cases

  // left-left insertion, balance factor > 1
  //     x (root)
  //    /
  //   y
  //  /
  // z
  // balance_factor > 1 means that height of the left subtree is
  // greater than that of the right subtree
  if (balance_factor > 1 && data < root->left->data) {
    printf("UNBALANCED SUBTREE DETECTED DUE TO LEFT-LEFT INSERTION\n");
    return rightRotate(root);
  }

  // right-right insertion, balance factor < -1
  // x (root)
  //  \
  //   y
  //   \
  //    z
  // balance_factor < 1 means that height of the left subtree is
  // less than that of the right subtree
  if (balance_factor < -1 && data > root->right->data) {
    printf("UNBALANCED SUBTREE DETECTED DUE TO RIGHT-RIGHT INSERTION\n");
    return leftRotate(root);
  }

  // left-right insertion, balance factor > 1
  //     x (root) <- right rotate here
  //    /
  //   y  <- left rotate here
  //    \
  //     z
  if (balance_factor > 1 && data > root->left->data) {
    printf("UNBALANCED SUBTREE DETECTED DUE TO LEFT-RIGHT INSERTION\n");
    root->left = leftRotate(root->left);
    return rightRotate(root);
  }

  // right-left insertion, balance factor < -1
  //  x (root) <- left rotate here
  //   \
  //    y <- right rotate here
  //   /
  //  z
  if (balance_factor < -1 && data < root->right->data) {
    printf("UNBALANCED SUBTREE DETECTED DUE TO RIGHT-LEFT INSERTION\n");
    root->right = rightRotate(root->right);
    return leftRotate(root);
  }

  return root;
}

// ── Print and free tree ──────────────────────────────────────────────────────
// in order tree walk
void inOrderWalk(const TreeNode *root) {
  if (root) {
    inOrderWalk(root->left);
    printf("%d ", root->data);
    inOrderWalk(root->right);
  }
}

// free tree
void freeTree(TreeNode *root) {
  if (root) {
    freeTree(root->left);
    free(root);
    freeTree(root->right);
  }
}

// ── main ─────────────────────────────────────────────────────────────────────
int main(int argc, char *argv[]) {
  int list[] = {33, 13, 53, 11, 21, 61, 8, 9};
  int size = sizeof(list) / sizeof(list[0]);

  TreeNode *root = NULL;
  for (int i = 0; i < size; i++) {
    root = insertNode(root, list[i]);
    // check balance factor on every insertion
    assert(getBalanceFactor(root) >= -1 && getBalanceFactor(root) <= 1);
  }

  printf("In-order: ");
  inOrderWalk(root);
  printf("\n");

  freeTree(root);
  return EXIT_SUCCESS;
}

File: Python/Data Structures/Trees/01_AVL_Trees/01_insertion.py
import sys


class Treenode(object):
    def __init__(self, data: int) -> None:
        self.data: int = data
        self.left = None
        self.right = None
        self.height: int = 1


class AVLTree:
    # get height of the current node
    def get_height(self, node):
        return 0 if not node else node.height

    def update_height(self, node):
        return 1 + max(self.get_height(node.left), self.get_height(node.right))

    def get_balance_factor(self, node):
        return (
            0
            if node is None
            else self.get_height(node.left) - self.get_height(node.right)
        )

    def left_rotate(self, x: Treenode):
        y = x.right
        t2 = y.left

        y.left = x
        x.right = t2

        self.update_height(x)
        self.update_height(y)

        return y

    def right_rotate(self, x: Treenode):
        y = x.left
        t2 = y.right

        y.right = x
        x.left = t2

        self.update_height(x)
        self.update_height(y)

        return y

    def insert_node(self, root: Treenode, key: int):
        if not root:
            return Treenode(key)

        if key > root.data:
            root.right = self.insert_node(root.right, key)
        else:
            root.left = self.insert_node(root.left, key)

        self.update_height(root)

        balance_factor = self.get_balance_factor(root)

        if balance_factor > 1:
            if key < root.left.data:
                return self.right_rotate(root)
            else:
                root.left = self.left_rotate(root.left)
                return self.right_rotate(root)

        if balance_factor < -1:
            if key > root.right.data:
                return self.left_rotate(root)
            else:
                root.right = self.right_rotate(root.right)
                return self.left_rotate(root)

        return root

    def in_order(self, root):
        if root:
            self.pre_order(root.left)
            print(f"{root.data}", end=" ")
            self.pre_order(root.right)

    # Print the tree
    def printHelper(self, currPtr, indent, last):
        if currPtr != None:
            sys.stdout.write(indent)
            if last:
                sys.stdout.write("R----")
                indent += "     "
            else:
                sys.stdout.write("L----")
                indent += "|    "
            print(currPtr.data)
            self.printHelper(currPtr.left, indent, False)
            self.printHelper(currPtr.right, indent, True)


myTree = AVLTree()
root = None
nums = [33, 13, 52, 9, 21, 61, 8, 11]
for num in nums:
    root = myTree.insert_node(root, num)
myTree.printHelper(root, "", True)
